### ISSUE TITLE
iOS & Android crash when animation occurs

### ISSUE THREAD

**Issue by alezhuk** [rotorgames/Rg.Plugins.Popup#94]
Hey! This plugin is great, but has issues with app crashing on both iOS and Android platforms. I've made some "stress tests" with open/close popups, and it's crashing with 99% of tests. I was testing on devices and emulators for latest different OS versions. Even when user is working with app for a while, we're getting lot's of crashes. They occur when popup try to be shown or closed. Currently version can't be used in production code. If there will be time, I would like to be contributor to provide some fixes, but have no time for that now.

Nice work, but need to be improved for sure.

Regards

**Comment by rotorgames** [comment]
@alezhuk Hi. Can you show stack trace or crash reports me? You say that you have a lot of crashes but I use this plugin in several production apps and it works fine.

**Comment by alezhuk** [comment]
@rotorgames sure, I'll prepare these for you. Thanks for the quick response

Regards

**Comment by rotorgames** [comment]
@alezhuk Ok. I will wait your response.

**Comment by rotorgames** [comment]
@alezhuk P.S. And add sample code of tests or examples.

**Comment by alezhuk** [comment]
@rotorgames emm.... this can be a problem, couse it is aka customer source. Some day, I'll write sample_stress_test_project

Regards

**Comment by rotorgames** [comment]
@alezhuk No problem, but if I don't see stack trace I can't fix this problem. Stack trace should be required.

**Comment by alezhuk** [comment]
@rotorgames sure, I'll post that here when I reach my work laptop. Thanks


Regards

**Comment by rotorgames** [comment]
@alezhuk Ok, thank you too.

**Comment by alezhuk** [comment]
@rotorgames here is for iOS. Got it after using my app for 5 mins.

Regards
[ios_stack_trace.cs.zip](https://github.com/rotorgames/Rg.Plugins.Popup/files/842720/ios_stack_trace.cs.zip)

**Comment by rotorgames** [comment]
@alezhuk Can you show stack trace from android bug report?

**Comment by rotorgames** [comment]
@alezhuk And write versions of iOS/Android, Xamarin.Forms. Xamarin.iOS/Xamarin.Android/Xamarin.VS.

**Comment by alezhuk** [comment]
@rotorgames sure, as I said, I'm out of the office. So will prepare all the info soon.

**Comment by rotorgames** [comment]
@alezhuk Sure. I see that is not only you have this problem. Please write iOS version which has this bug.

**Comment by alezhuk** [comment]
@rotorgames for now, I see that solution is to disable animate on pop

Regards

**Comment by rotorgames** [comment]
@alezhuk Yes but it is not fix bug. Maybe this problem exists only in the new versions of OS. I think so because similar stack trace I begin to get recently.

**Comment by alezhuk** [comment]
@rotorgames you can easily or not easily reproduce that crash (not sure for OS versions for now, need to test that stuff), just show/close popup inside the loop for example 100 times.

**Comment by rotorgames** [comment]
@alezhuk Did you wait the `await` method in the loop? Did you open single instance of popup page or you created new instance in every loop?

**Comment by alezhuk** [comment]
@rotorgames yes, with await. I've tried both as single instance and new instance, the result is the same. Anyway, it's some kind of heisenbug ;)

**Comment by rotorgames** [comment]
@alezhuk Did you close popup page before opening new popup page or you created 100 pages on display?

**Comment by alezhuk** [comment]
@rotorgames in my tests I was also using aka lazy init:
`private MyClass _myProperty;

public MyClass MyProperty
{
    get
    {
        if (_myProperty == null)
        {
            _myProperty = new MyClass();
        }
        return _myProperty;
    }
}`

**Comment by rotorgames** [comment]
@alezhuk Is `MyClass` popup page?

**Comment by alezhuk** [comment]
@rotorgames let me use no abstractions ;)
```
private PopupLoadingPage popupLoadingPage;
        public PopupLoadingPage PopupLoadingPage
        {
            get
            {
                return popupLoadingPage ??
                    (popupLoadingPage = new PopupLoadingPage());
            }
        }
```

Here, the **PopupLoadingPage** is:
```
public partial class PopupLoadingPage : Rg.Plugins.Popup.Pages.PopupPage
    {
        public PopupLoadingPage()
        {
            InitializeComponent();
        }
    }
```
which has only **ActivityIndicator** inside and **CloseWhenBackgroundIsClicked="false"**

**Comment by rotorgames** [comment]
@alezhuk also give me your for when you open and close popup

**Comment by alezhuk** [comment]
```
public async Task ShowLoadingAsync()
        {
            await PopupNavigation.PushAsync(PopupLoadingPage);
        }
```

**Comment by rotorgames** [comment]
@alezhuk Ok. Thank you for quick responses. I will test it tomorrow.

**Comment by alezhuk** [comment]
@rotorgames see the updated version

Regards

**Comment by alezhuk** [comment]
@rotorgames  Also, I'm using another method from my interface to dismiss any popup:
```
public async Task ClosePopupAsync()
        {
            if (PopupNavigation.PopupStack.Count > uint.MinValue)
                await PopupNavigation.PopAsync();
        }
```

**Comment by alezhuk** [comment]
@rotorgames I believe these all might help you to investigate.

Regards

**Comment by rotorgames** [comment]
But why? You can use 'PopAllAsync'

**Comment by alezhuk** [comment]
@rotorgames In my prod, no more than one popup is displayed. I like mobile design patterns, so in any cool app (except games) there must be not more than one popup page at time. And of cource this only my Opinion :)

**Comment by rotorgames** [comment]
@alezhuk Hi. I tested it and I didn't get any crashes. I did tests on iOS Simulator 10.2. 

I used this code for stress test 

```csharp
private async void OnOpenPupup(object sender, EventArgs e)
{
    for (int i = 0; i < 100; i++)
    {
        await Navigation.PushPopupAsync(new LoginPopupPage());
        await Navigation.PopPopupAsync();
    }
}
```
also I used:
```csharp
private async void OnOpenPupup(object sender, EventArgs e)
{
    for (int i = 0; i < 100; i++)
    {
        await Navigation.PushPopupAsync(_loginPopup); // single instance
        await Navigation.PopPopupAsync();
    }
}
```

Can you show your xaml code which did you use with tests?

**Comment by alezhuk** [comment]
iOS Version 8.4.1
Xamarin.iOS Version: 10.3.1.8
[ios_crash.cs.zip](https://github.com/rotorgames/Rg.Plugins.Popup/files/850724/ios_crash.cs.zip)

**Comment by alezhuk** [comment]
I have the same tests implemented, but it is strange that you have got no crash. Also I've tried to disable animations, but no luck for now.

**Comment by rotorgames** [comment]
@alezhuk I tested on iOS 10.2. maybe for this reason I could not get any errors.

**Comment by rotorgames** [comment]
@alezhuk P.S. Soon I publish new version v1.2.0 where iOS work will be changed. But I should wait when Xamarin.Forms v2.3.4 will be published

**Comment by rotorgames** [comment]
@alezhuk P.P.S I managed to reproduce this error on iOS 8.4. I will fix it

**Comment by alezhuk** [comment]
@rotorgames This is great! Are you able to reproduce the same issue for Android 5+ version?

**Comment by rotorgames** [comment]
@alezhuk I think that Popup plugin has memory leaks and it is source of problem. I will review this plugin. It need much time.

**Comment by anudr01d** [comment]
@rotorgames Any temporary fixes available? I'm seeing a lot of crashes with or without animations set to true.

**Comment by rotorgames** [comment]
@anudr01d I should wait Xamarin.Forms 2.3.4. Only then I will be able to try to fix this bug.

**Comment by alezhuk** [comment]
FYI https://blog.xamarin.com/announcing-xamarin-forms-stable-release-2-3-4/

**Comment by rotorgames** [comment]
@alezhuk Yes. I have seen it. I'm working on it. New release will be soon.

**Comment by ArtemKumeisha** [comment]
Hi. I have problems with animation too. My PopUpPage : https://gist.github.com/ArtemKumeisha/7c851adce3131a7548366f97bd5a5b9b
When I do several times await PopupNavigation.PushAsync(new DisplayPopUpPage(isRunning, message));/await PopupNavigation.PopAsync();  app  get fatal error : https://gist.github.com/ArtemKumeisha/e0d0e3cc4d8e89a692b473e3bba8b13c
This can happen after 10 times, and maybe even after a second run. Thanks!!

**Comment by rotorgames** [comment]
@ArtemKumeisha @alezhuk @anudr01d This issue was fixed in v1.1.0-pre1. Please test it and tell me about your tests. Thank you.
